### DISCUSSION TITLE
Embedding app into react application

### DISCUSSION THREAD

**Discussion by KLR2002** [AaronDavidNewman/Smoosic#104] [Q&A]
Hi!

Great app, you're doing a good job!

I am also currently working on a music-related application (choir management and OMR). I thought it would be good to have something to edit/review sheet music compositions and came across your app. The only thing is that I don't really know how I can build it into my app in React. I haven't found information on what things I can import and how I can use them. Can you help me out?

**Comment by AaronDavidNewman** [comment]
Thanks, glad you enjoy it.  

Smoosic renders the music in an SVG div, so you can use the library to render the music yourself in your own application.  There is also an 'application' mode that manages browser events like mouse etc. used in the full application.   You can either customize it or replace it with your own logic.

If you look in the Demos repository, there are a number of demos of Smoosic, both in application and library form.  If you use just the library (look at asyncDemo.html/asyncDemo.ts), you can see rendering a score and managing the selections programmatically.  There are also demos there for importing/exporting different file types, mostly musicXML and the native Smoosic format would apply to your situation, and creating custom UI extensions.

In the src/application directory, you can see both the eventHandler.ts and keyCommands.ts to see how we handle the low-level browser events.  The events either go to render/sui/tracker.ts, which manages the 'cursor' and selections, or to the keyCommands, which usually calls something in render/sui/scoreViewOperations.ts where most of the score manipulation is done.

**Reply by KLR2002** [reply]
Thanks for the answer. I managed to modify the UI a bit to my needs. I have a problem though, I don't know what part of the code is responsible for loading the files. What I want to achieve is that when the application starts, the MusicXML file is loaded and displayed instead of these 4 pauses. I tried entering the path to this file into either initialScore or remoteScore, but this did not work

**Comment by AaronDavidNewman** [comment]
Maybe look at SuiLoadMxmlDialog in src/ui/dialogs/fileDialog.ts.  

Once you have the XML, the code to get it into the application and rendered is like this (in the same file, look at SuiXmlLoadAdapter).  Assuming 'xmlFile' has your xml file in string form:

``` typescript
      const parser = new DOMParser();
      const xml = parser.parseFromString(xmlFile, 'text/xml');
      const score = XmlToSmo.convert(xml);
      await view.changeScore(score);
```
`view` is the SuiScoreViewOperations object which you can get from the application object.

The Finale .mxl file is a zipped file that contains the xml string, it _should_ work with either that or the text xml file.  So maybe to test your code, you could try to just put an xml string directly in your code, and once that works, figure out how to get at it through the fileio code, or your own adapted version of it.  It's quite possible there is metadata in the mxl files that I don't handle properly.
